// rsuite/cjs/Message/Message.js
'use client';
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");
exports.__esModule = true;
exports.default = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _taggedTemplateLiteralLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteralLoose"));
var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));
var _react = _interopRequireWildcard(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _utils = require("../utils");
var _propTypes2 = require("../internals/propTypes");
var _CloseButton = _interopRequireDefault(require("../internals/CloseButton"));
var _ToastContext = _interopRequireDefault(require("../toaster/ToastContext"));
var _templateObject, _templateObject2, _templateObject3, _templateObject4, _templateObject5;
/**
 * The `Message` component is used to display important messages to users.
 * @see https://rsuitejs.com/components/message
 */
var Message = /*#__PURE__*/_react.default.forwardRef(function (props, ref) {
  var _withClassPrefix;
  var _props$as = props.as,
    Component = _props$as === void 0 ? 'div' : _props$as,
    bordered = props.bordered,
    centered = props.centered,
    className = props.className,
    _props$classPrefix = props.classPrefix,
    classPrefix = _props$classPrefix === void 0 ? 'message' : _props$classPrefix,
    children = props.children,
    closable = props.closable,
    _props$duration = props.duration,
    duration = _props$duration === void 0 ? 2000 : _props$duration,
    full = props.full,
    header = props.header,
    _props$type = props.type,
    type = _props$type === void 0 ? 'info' : _props$type,
    showIcon = props.showIcon,
    onClose = props.onClose,
    rest = (0, _objectWithoutPropertiesLoose2.default)(props, ["as", "bordered", "centered", "className", "classPrefix", "children", "closable", "duration", "full", "header", "type", "showIcon", "onClose"]);
  var _useState = (0, _react.useState)('show'),
    display = _useState[0],
    setDisplay = _useState[1];
  var _useClassNames = (0, _utils.useClassNames)(classPrefix),
    withClassPrefix = _useClassNames.withClassPrefix,
    merge = _useClassNames.merge,
    prefix = _useClassNames.prefix;
  var isMounted = (0, _utils.useIsMounted)();
  var _useContext = (0, _react.useContext)(_ToastContext.default),
    usedToaster = _useContext.usedToaster; // Timed close message
  var _useTimeout = (0, _utils.useTimeout)(onClose, duration, usedToaster && duration > 0),
    clear = _useTimeout.clear;
  var handleClose = (0, _utils.useEventCallback)(function (event) {
    setDisplay('hiding');
    onClose === null || onClose === void 0 ? void 0 : onClose(event);
    clear();
    setTimeout(function () {
      if (isMounted()) {
        setDisplay('hide');
      }
    }, 1000);
  });
  if (display === 'hide') {
    return null;
  }
  var classes = merge(className, withClassPrefix(type, display, (_withClassPrefix = {
    full: full,
    bordered: bordered,
    centered: centered
  }, _withClassPrefix['has-title'] = header, _withClassPrefix['has-icon'] = showIcon, _withClassPrefix)));
  return /*#__PURE__*/_react.default.createElement(Component, (0, _extends2.default)({
    role: "alert"
  }, rest, {
    ref: ref,
    className: classes
  }), /*#__PURE__*/_react.default.createElement("div", {
    className: prefix(_templateObject || (_templateObject = (0, _taggedTemplateLiteralLoose2.default)(["container"])))
  }, closable && /*#__PURE__*/_react.default.createElement(_CloseButton.default, {
    onClick: handleClose
  }), showIcon && /*#__PURE__*/_react.default.createElement("div", {
    className: prefix(_templateObject2 || (_templateObject2 = (0, _taggedTemplateLiteralLoose2.default)(["icon"])))
  }, _utils.MESSAGE_STATUS_ICONS[type]), /*#__PURE__*/_react.default.createElement("div", {
    className: prefix(_templateObject3 || (_templateObject3 = (0, _taggedTemplateLiteralLoose2.default)(["content"])))
  }, header && /*#__PURE__*/_react.default.createElement("div", {
    className: prefix(_templateObject4 || (_templateObject4 = (0, _taggedTemplateLiteralLoose2.default)(["header"])))
  }, header), children && /*#__PURE__*/_react.default.createElement("div", {
    className: prefix(_templateObject5 || (_templateObject5 = (0, _taggedTemplateLiteralLoose2.default)(["body"])))
  }, children))));
});
Message.displayName = 'Message';
Message.propTypes = {
  bordered: _propTypes.default.bool,
  centered: _propTypes.default.bool,
  closable: _propTypes.default.bool,
  className: _propTypes.default.string,
  classPrefix: _propTypes.default.string,
  description: _propTypes.default.node,
  full: _propTypes.default.bool,
  onClose: _propTypes.default.func,
  showIcon: _propTypes.default.bool,
  title: _propTypes.default.node,
  type: (0, _propTypes2.oneOf)(_utils.STATUS)
};
var _default = Message;
exports.default = _default;
